plugins/techdocs-react/src/component.tsx

```diff
@@ -0,0 +1,227 @@
+ * @public
+ */
+export const useShadowDomStylesLoading = (element: Element | null) => {
+  const [loading, setLoading] = useState(false);
+
+  useEffect(() => {
+    if (!element) return () => {};
+
+    setLoading(true);
+
+    const style = (element as HTMLElement).style;
+
+    style.setProperty('opacity', '0');
+
+    const handleLoad = () => {
+      // As styles loading can be fast, a delay is used to display a progress bar
+      // and give the user time to see it before displaying the page content.
+      setTimeout(() => {
+        setLoading(false);
+        style.setProperty('opacity', '1');
+      }, 500);
+    };
+
+    element.addEventListener(SHADOW_DOM_STYLE_LOAD_EVENT, handleLoad);
+
+    return () => {
+      element.removeEventListener(SHADOW_DOM_STYLE_LOAD_EVENT, handleLoad);
+    };
+  }, [element]);
+
+  return loading;
+};
+
+/**
+ * Props for {@link TechDocsShadowDom}.
+ *
+ * @remarks
+ * If you want to use portals to render Material UI components in the Shadow DOM,
+ * you must render these portals as children because this component wraps its children in a Material UI StylesProvider
+ * to ensure that Material UI styles are applied.
+ *
+ * @public
+ */
+export type TechDocsShadowDomProps = PropsWithChildren<{
+  /**
+   * Element tree that is appended to ShadowRoot.
+   */
+  element: Element;
+  /**
+   * Callback called when the element tree is appended in ShadowRoot.
+   */
+  onAppend?: (shadowRoot: ShadowRoot) => void;
+}>;
+
+/**
+ * Renders a tree of elements in a Shadow DOM.
+ *
+ * @remarks
+ * Centers the styles loaded event to avoid having multiple locations setting the opacity style in Shadow DOM causing the screen to flash multiple times,
+ * so if you want to know when Shadow DOM styles are computed, you can listen for the "TECH_DOCS_SHADOW_DOM_STYLE_LOAD" event dispatched by the element tree.
+ *
+ * @example
+ * Here is an example using this component and also listening for styles loaded event:
+ *```jsx
+ * import {
+ *   TechDocsShadowDom,
+ *   SHADOW_DOM_STYLE_LOAD_EVENT,
+ * } from '@backstage/plugin-techdocs-react';
+ *
+ * export const TechDocsReaderPageContent = ({ entity }: TechDocsReaderPageContentProps) => {
+ *   // ...
+ *   const dom = useTechDocsReaderDom(entity);
+ *
+ *   useEffect(() => {
+ *     const updateSidebarPosition = () => {
+ *       // ...
+ *     };
+ *     dom?.addEventListener(SHADOW_DOM_STYLE_LOAD_EVENT, updateSidebarPosition);
+ *     return () => {
+ *       dom?.removeEventListener(SHADOW_DOM_STYLE_LOAD_EVENT, updateSidebarPosition);
+ *     };
+ *   }, [dom]);
+ *
+ *   const handleDomAppend = useCallback(
+ *     (newShadowRoot: ShadowRoot) => {
+ *       setShadowRoot(newShadowRoot);
+ *     },
+ *     [setShadowRoot],
+ *   );
+ *
+ *   return <TechDocsShadowDom element={dom} onAppend={handleDomAppend} />;
+ * };
+ * ```
+ *
+ * @param props - see {@link TechDocsShadowDomProps}.
+ * @public
+ */
+export const TechDocsShadowDom = ({
```

Comment:
I was confused because we also have `hooks` and `context` files that contain code related to shadow dom and they have generic filenames so I followed the pattern... Do you think it would be nice to group them all in `shadow-dom ` folder ?

Reply:
I see.. hmm yeah that would probably clarify. But lets do that moving separetly then and keep as it is now!